### PR TITLE
Fix excessive SettableFuture retention

### PR DESCRIPTION
**Pull request by cberner**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java [hunk 1]
```diff
@@ -84,6 +87,22 @@ public T get()
 
             SettableFuture<T> futureStateChange = SettableFuture.create();
             futureStateChanges.add(futureStateChange);
+            Futures.addCallback(futureStateChange, new FutureCallback<T>() {
+                @Override
+                public void onSuccess(T result)
+                {
+                    // no-op
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java]: Maybe add a comment explaining why we don't need to do it here, too

File: presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java [hunk 2]
```diff
@@ -186,7 +205,7 @@ private void fireStateChanged(T newState, List<SettableFuture<T>> futureStateCha
         checkState(!Thread.holdsLock(this), "Can not fire state change event while holding a lock on this");
 
         executor.execute(() -> {
-            checkState(!Thread.holdsLock(StateMachine.this), "Can not notify while holding a lock on this");
+            checkState(!Thread.holdsLock(this), "Can not notify while holding a lock on this");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java]: Is this intentional? "this" in this context is the lambda

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java]: Scratch that... this is correct. It used to be a Runnable, so "this" was the runnable. With lambdas, "this" is the enclosing class.

File: presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java [hunk 3]
```diff
@@ -52,7 +55,7 @@
     private final List<StateChangeListener<T>> stateChangeListeners = new ArrayList<>();
 
     @GuardedBy("this")
-    private final List<SettableFuture<T>> futureStateChanges = new ArrayList<>();
+    private final Set<SettableFuture<T>> futureStateChanges = Sets.newIdentityHashSet();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/StateMachine.java]: Static import

**Comment by dain** [comment]
looks good

**Pull request merged**